frontend/lib/src/components/elements/Popover/Popover.tsx

```diff
@@ -0,0 +1,142 @@
+ */
+
+import React, { ReactElement } from "react"
+
+import { useTheme } from "@emotion/react"
+import { ExpandMore, ExpandLess } from "@emotion-icons/material-outlined"
+
+import { hasLightBackgroundColor } from "@streamlit/lib/src/theme"
+import { StyledIcon } from "@streamlit/lib/src/components/shared/Icon"
+import { Block as BlockProto } from "@streamlit/lib/src/proto"
+import BaseButton, {
+  BaseButtonKind,
+  BaseButtonSize,
+} from "@streamlit/lib/src/components/shared/BaseButton"
+import StreamlitMarkdown from "@streamlit/lib/src/components/shared/StreamlitMarkdown"
+import IsSidebarContext from "@streamlit/lib/src/components/core/IsSidebarContext"
+
+import { Popover as UIPopover, TRIGGER_TYPE, PLACEMENT } from "baseui/popover"
+
+import { StyledPopoverButtonIcon } from "./styled-components"
+
+export interface PopoverProps {
+  element: BlockProto.Popover
+  empty: boolean
+}
+
+const Popover: React.FC<PopoverProps> = ({
+  element,
+  empty,
+  children,
+}): ReactElement => {
+  const [open, setOpen] = React.useState(false)
+  const isInSidebar = React.useContext(IsSidebarContext)
+  const theme = useTheme()
+
+  const lightBackground = hasLightBackgroundColor(theme)
+
+  return (
+    <div data-testid="stPopover">
+      <UIPopover
+        triggerType={TRIGGER_TYPE.click}
+        placement={PLACEMENT.bottomLeft}
+        content={() => children}
+        isOpen={open}
+        onClickOutside={() => setOpen(false)}
+        onEsc={() => setOpen(false)}
+        ignoreBoundary={isInSidebar}
+        overrides={{
+          Body: {
+            props: {
+              "data-testid": "stPopoverBody",
+            },
+            style: () => ({
+              marginRight: theme.spacing.lg,
+              marginBottom: theme.spacing.lg,
+              maxHeight: "70vh",
+              overflow: "auto",
+              maxWidth: "80vw",
+              minWidth: "20rem",
+
+              paddingRight: "calc(1em - 1px)", // 1px to account for border.
+              paddingLeft: "calc(1em - 1px)",
+              paddingBottom: "calc(1em - 1px)",
+              paddingTop: "calc(1em - 1px)",
+
+              borderLeftWidth: "1px",
+              borderRightWidth: "1px",
+              borderTopWidth: "1px",
+              borderBottomWidth: "1px",
+
+              borderLeftRadius: theme.radii.lg,
+              borderRightRadius: theme.radii.lg,
+              borderTopRadius: theme.radii.lg,
+              borderBottomRadius: theme.radii.lg,
+
+              borderLeftStyle: "solid",
+              borderRightStyle: "solid",
+              borderTopStyle: "solid",
+              borderBottomStyle: "solid",
+
+              borderLeftColor: theme.colors.fadedText10,
+              borderRightColor: theme.colors.fadedText10,
+              borderTopColor: theme.colors.fadedText10,
+              borderBottomColor: theme.colors.fadedText10,
+
+              boxShadow: lightBackground
+                ? "0px 4px 16px rgba(0, 0, 0, 0.16)"
```

Comment:
Is the boxShadow used in other components as well? Would it make sense to add it to the `theme` and reference here?

Reply:
Its based on the toast shadow, but we also have some other types of shadows (e..g for menus, tooltips, toolbar, ..) so its more of a case by case basis. But I think we should eventually move more of these things into theming. But I think this needs to be more of a dedicated thought out project.

Reply:
I also added it to the tech debt item [here](https://www.notion.so/snowflake-corp/Refactor-theming-aspects-09278d1c715f4e8ea5fc1a0019911ab2).